test/test_qa_utils.py

```diff
@@ -1048,3 +1115,256 @@ def test_mock_keys_not_implemented():
+        assert printed.last == "<CustomizableProperty FAVORITE_SONG>"
+    with printed_output() as printed:
+        with raises_regexp(UncustomizedInstance, "Attempt to access field FAVORITE_SONG"
+                                                 " from class test.test_qa_utils.SampleClass2."
+                                                 " It was expected to be given a custom value in a subclass:"
+                                                 " the string name of a song."):
+            PRINT(getattr_customized(SampleClass2, 'FAVORITE_SONG'))
+        assert printed.last is None
+
+
+def test_raises_regexp():
+
+    class MyRuntimeError(RuntimeError):
+        pass
+
+    with raises_regexp(RuntimeError, "This.*test!"):
+        raise RuntimeError("This is a test!")
+
+    with raises_regexp(RuntimeError, "This.*test!"):
+        raise MyRuntimeError("This is a test!")
+
+    with pytest.raises(AssertionError):
+        # This will fail because the inner error has a period, not an exclamation mark, terminating it.
+        # That will cause it to raise an AssertionError instead.
+        with raises_regexp(RuntimeError, "This.*test!"):
+            raise MyRuntimeError("This is a test.")
+
+    with pytest.raises(Exception):
+        # This will fail because the inner error is a KeyError, not a RuntimeError.
+        # I WISH this would raise AssertionError, but pytest lets the KeyError through.
+        # I am not sure that's the same as what unittest does in this case but it will
+        # suffice for now. -kmp 6-Oct-2020
+        with raises_regexp(RuntimeError, "This.*test!"):
+            raise KeyError('This is a test!')
+
+
+def test_version_checker_no_changelog():
+
+    class MyVersionChecker(VersionChecker):
+        PYPROJECT = os.path.join(os.path.dirname(__file__), "../pyproject.toml")
+        CHANGELOG = None
+
+    MyVersionChecker.check_version()
+
+
+def test_version_checker_use_dcicutils_changelog():
+
+    class MyVersionChecker(VersionChecker):
+        PYPROJECT = os.path.join(os.path.dirname(__file__), "../pyproject.toml")
+        CHANGELOG = os.path.join(os.path.dirname(__file__), "../CHANGELOG.md")
+
+    MyVersionChecker.check_version()
+
+
+def test_version_checker_with_missing_changelog():
+
+    path_exists = os.path.exists
+
+    def mocked_exists(filename):
+        if filename.endswith("CHANGELOG.md"):
+            print("Faking that %s does not exist." % filename)
+            return False
+        else:
+            return path_exists(filename)
+
+    with mock.patch("os.path.exists", mocked_exists):
+
+        class MyVersionChecker(VersionChecker):
+
+            PYPROJECT = os.path.join(os.path.dirname(__file__), "../pyproject.toml")
+            CHANGELOG = os.path.join(os.path.dirname(__file__), "../CHANGELOG.md")
+
+        with pytest.raises(AssertionError):
+            MyVersionChecker.check_version()  # The version history will be missing because of mocking.
```

Comment:
I'm mixed on whether I think things like this should throw assertion error, despite it being used in QA.

Reply:
What would you use instead?